src/solver_routines.f90

```diff
@@ -18679,4 +19081,182 @@ SUBROUTINE SOLVER_NONLINEAR_MONITOR_PETSC(SNES,ITS,NORM,CTX,ERR)
+  NULLIFY(MODELS_VARIABLE)
+
+  IF(ASSOCIATED(CTX)) THEN
+    DOF_NUMBER = CTX%DOF_NUMBER
+    IF(ASSOCIATED(CTX%SOLVER)) THEN
+      CELLML_EQUATIONS=>CTX%SOLVER%CELLML_EQUATIONS
+      IF(ASSOCIATED(CELLML_EQUATIONS)) THEN
+        DO cellml_idx=1,CELLML_EQUATIONS%NUMBER_OF_CELLML_ENVIRONMENTS
+          CELLML_ENVIRONMENT=>CELLML_EQUATIONS%CELLML_ENVIRONMENTS(cellml_idx)%PTR
+           MAX_NUMBER_STATES=CELLML_ENVIRONMENT%MAXIMUM_NUMBER_OF_STATE
+           MAX_NUMBER_INTERMEDIATES=CELLML_ENVIRONMENT%MAXIMUM_NUMBER_OF_INTERMEDIATE
+           MAX_NUMBER_PARAMETERS=CELLML_ENVIRONMENT%MAXIMUM_NUMBER_OF_PARAMETERS
+          !Make sure CellML fields have been updated to the current value of any mapped fields
+
+          IF(ASSOCIATED(CELLML_ENVIRONMENT)) THEN                  
+            CELLMLMODELSFIELD=>CELLML_ENVIRONMENT%MODELS_FIELD
+            !Get the parameters information if this environment has any.
+            IF(ASSOCIATED(CELLML_ENVIRONMENT%PARAMETERS_FIELD)) THEN
+              PARAMETERS_FIELD=>CELLML_ENVIRONMENT%PARAMETERS_FIELD%PARAMETERS_FIELD
+              IF(ASSOCIATED(PARAMETERS_FIELD)) THEN
+                CALL FIELD_PARAMETER_SET_DATA_GET(PARAMETERS_FIELD, &
+                  & FIELD_U_VARIABLE_TYPE,FIELD_VALUES_SET_TYPE, &
+                  & PARAMETERS_DATA,ERR,ERROR,*999)
+              ENDIF
+            ENDIF
+
+            !Get the intermediate information if this environment has any.
+            IF(ASSOCIATED(CELLML_ENVIRONMENT%INTERMEDIATE_FIELD)) THEN
+              INTERMEDIATE_FIELD=>CELLML_ENVIRONMENT%INTERMEDIATE_FIELD%INTERMEDIATE_FIELD
+              IF(ASSOCIATED(INTERMEDIATE_FIELD)) THEN
+                 CALL FIELD_PARAMETER_SET_DATA_GET(INTERMEDIATE_FIELD, &
+                   & FIELD_U_VARIABLE_TYPE,FIELD_VALUES_SET_TYPE, &
+                   & INTERMEDIATES_DATA,ERR,ERROR,*999)                            
+              ENDIF
+            ENDIF!associated intermediate
+            IF(ASSOCIATED(CELLMLMODELSFIELD)) THEN
+              MODELS_FIELD=>CELLMLMODELSFIELD%MODELS_FIELD
+
+              IF(ASSOCIATED(MODELS_FIELD)) THEN
+                CALL FIELD_VARIABLE_GET(MODELS_FIELD,FIELD_U_VARIABLE_TYPE,MODELS_VARIABLE,ERR,ERROR,*999)
+                CALL FIELD_DOF_ORDER_TYPE_GET(MODELS_FIELD,FIELD_U_VARIABLE_TYPE,DOF_ORDER_TYPE,ERR,ERROR,*999)
+                CALL FIELD_PARAMETER_SET_DATA_GET(MODELS_FIELD,FIELD_U_VARIABLE_TYPE,FIELD_VALUES_SET_TYPE, &
+                  & MODELS_DATA,ERR,ERROR,*999)
+                model_idx = MODELS_DATA(DOF_NUMBER)
+                IF(ASSOCIATED(CELLML_ENVIRONMENT%MODELS(model_idx)%PTR)) THEN
+                  MODEL=>CELLML_ENVIRONMENT%MODELS(model_idx)%PTR
+                  NUMBER_STATES = MODEL%NUMBER_OF_STATE
+                  NUMBER_INTERMEDIATES=MODEL%NUMBER_OF_INTERMEDIATE
+                  NUMBER_PARAMETERS=MODEL%NUMBER_OF_PARAMETERS
+                  !Get the states from PETSc
+                  ALLOCATE(STATES(NUMBER_STATES),STAT=ERR)
```

Comment:
Allocating inside computational loops should be a last, last resort due to performance implications.